tutorials/shaders/shader_reference/shader_functions.rst

```diff
@@ -0,0 +1,5273 @@
+    | vec_type        | float, vec2, vec3, or vec4                          | genType                  |
+    +-----------------+-----------------------------------------------------+--------------------------+
+    | vec_int_type    | int, ivec2, ivec3, or ivec4                         | genIType                 |
+    +-----------------+-----------------------------------------------------+--------------------------+
+    | vec_uint_type   | uint, uvec2, uvec3, or uvec4                        | genUType                 |
+    +-----------------+-----------------------------------------------------+--------------------------+
+    | vec_bool_type   | bool, bvec2, bvec3, or bvec4                        | genBType                 |
+    +-----------------+-----------------------------------------------------+--------------------------+
+    | mat_type        | mat2, mat3, or mat4                                 | mat                      |
+    +-----------------+-----------------------------------------------------+--------------------------+
+    | gvec4_type      | vec4, ivec4, or uvec4                               | gvec4                    |
+    +-----------------+-----------------------------------------------------+--------------------------+
+    | gsampler2D      | sampler2D, isampler2D, or uSampler2D                | gsampler2D               |
+    +-----------------+-----------------------------------------------------+--------------------------+
+    | gsampler2DArray | sampler2DArray, isampler2DArray, or uSampler2DArray | gsampler2DArray          |
+    +-----------------+-----------------------------------------------------+--------------------------+
+    | gsampler3D      | sampler3D, isampler3D, or uSampler3D                | gsampler3D               |
+    +-----------------+-----------------------------------------------------+--------------------------+
+
+    If  any of these are specified for multiple parameters, they must all be the same type unless otherwise noted.
+
+.. note::
+    Most operations on vectors (multiplication, division, etc) are performed component-wise.
+    For example, the operation ``vec2(3, 4) * vec2(10, 20)`` would result in ``vec2(3 * 10, 4 * 20)``.
+    or the operation ``min(vec2(3, 4), vec2(1, 8))`` would result in ``vec2(min(3, 1), min(4, 8))``.
+
+    The `GLSL Language Specification <http://www.opengl.org/registry/doc/GLSLangSpec.4.30.6.pdf>` says under section 5.10 Vector and Matrix Operations:
+
+        With a few exceptions, operations are component-wise. Usually, when an operator operates on a
+        vector or matrix, it is operating independently on each component of the vector or matrix,
+        in a component-wise fashion. [...] The exceptions are matrix multiplied by vector,
+        vector multiplied by matrix, and matrix multiplied by matrix. These do not operate component-wise,
+        but rather perform the correct linear algebraic multiply.
```

Comment:
This is probably better suited in the section on data types rather than here.

https://docs.godotengine.org/en/latest/tutorials/shaders/shader_reference/shading_language.html#data-types

Reply:
For now I copied the note into the main shading language page as well. If that works for you: great. If not, let me know what you would like me to try next.

Reply:
I think what makes this awkward is the entire section is talking about _operations_ not _functions_. GLSL isn't a language where you can override operators or customize them in any way. They are, for all intents and purposes, parts of the data type, not custom functions. 

But, reflecting on your comments, perhaps the best spot is in the section dedicated to operators https://docs.godotengine.org/en/latest/tutorials/shaders/shader_reference/shading_language.html#operators

Reply:
Ah, yes I can see how that would be a little offsetting. Putting it in the operators section is a good idea.

However, as far as I can tell this principle also applies to all the built in functions that have 2 vector parameters. This is why I believe that some version of this note still belongs in the built-in-functions section. I have three thoughts on ways to make this more clear:

1. I could change the wording of the existing note. Perhaps change all instances of the word 'operator' to 'function': "most functions that accept 2 vector parameters work component-wise..." Could also shorten the note a bit by not mentioning operators at all.
2. Remove the existing note and make a shorter one that is applied to each function that accepts two vector parameters.  Since we have a lot of overloads that are listed together, might have to call out specific method signatures in the note, but this would beore explicitly clear than a single generic note at the top.
3. Remove this note and trust that people writing shaders will learn this principle elsewhere or otherwise figure it out on their own. Moving the note to other pages more or less counts too, since this applies to many functions on this page and they may not think to look for such notes on other pages. Though, I agree that it is also applicable on other pages.
4. Move the content of this note to a new "component-wise operations" header on the shading language page and then add very brief notes ("component-wise function") that hyperlink to that section wherever they are applicable.

Reply:
@clayjohn Thoughts?

Reply:
My initial thought would be to add a single adjective "per-component" or "component-wise" to any relevant functions.  I suppose you could even put "Component-wise." as a whole sentence if the adjective doesn't fit naturally. Probably still keep a section somewhere else that has a more in-depth explanation. This accomplishes the goal of minimal text, while having each function be more or less self-documenting, *even if linked to on its own, to a new shader user*.

Which functions should it be added to? Either:
- Any functions that take in a `vec_type` parameter and does per-component operations. The problem is there are a *lot* of these.
- Any functions that take in *multiple* `vec_type` parameters and does per-component operations. Maybe the single parameter functions obviously work component-wise, since they take a vec4 and return a vec4.
---
I also like option 1. I would mention "operators and functions", and use an example of a function in addition to an operation, maybe something simple like `abs()` or `max()`.
I like the idea of option 4, in conjunction with a "component-wise" as a single word link in each function description, if that doesn't bloat the links too much.

Reply:
> Which functions should it be added to? Either:
> 
>  - Any functions that take in a vec_type parameter and does per-component operations. The problem is there are a lot of these.
>  - Any functions that take in multiple vec_type parameters and does per-component operations. Maybe the single parameter functions obviously work component-wise, since they take a vec4 and return a vec4.

Yeah, there's more than 100 functions listed here, almost all of them deal with vectors. I don't really have time right now to confirm which ones operate componentwise (almost all of them) and which ones actually perform an algebraic function. So I'm going to assume that its obvious that a single parameter function works component-wise.

Reply:
Also, I'm working on it now.

Reply:
Well, I discovered that the official [OpenGL ES 4.0 specification](https://registry.khronos.org/OpenGL/specs/gl/GLSLangSpec.4.30.pdf) is very good about specifying which functions perform component-wise. So I was actually able to just put a note on each function that does so. Will have a new commit up shortly.

Reply:
ok, @tetrapod00  and @clayjohn, just pushed up a commit that puts a Hyperlink above the description of each relevant function.

Reply:
I like this solution.
For entries that have a standalone "componentwise" link, I'd put the link after the description and before the parameter table, so the more relevant info about what the function actually does comes first.

I would wait until clayjohn approves this solution in general before making any more tedious changes, though.